From: Jan Kara <jack@suse.cz>
Subject: Fix heartbeat block writing
Patch-mainline: Never (has different code)
References: 300730

vec_len was computed wrong when number of nodes was smaller than fits into
the page and when vec_start was non-zero. This could lead to wrong update
of heartbeat information and bogus error messages or even assertion failures.

Signed-off-by: Jan Kara <jack@suse.cz>

--- a/fs/ocfs2/cluster/disk_heartbeat.c	2007-10-10 15:28:00.000000000 +0200
+++ b/fs/ocfs2/cluster/disk_heartbeat.c	2007-10-10 16:05:18.000000000 +0200
@@ -245,11 +245,9 @@ static struct bio *o2hb_setup_one_bio(st
 		current_page = cs / spp;
 		page = reg->hr_slot_data[current_page];
 
-		vec_len = min(PAGE_CACHE_SIZE,
+		vec_len = min(PAGE_CACHE_SIZE - vec_start,
 			      (max_slots-cs) * (PAGE_CACHE_SIZE/spp) );
 
-		vec_len -=  vec_start;
-
 		mlog(ML_HB_BIO, "page %d, vec_len = %u, vec_start = %u\n",
 		     current_page, vec_len, vec_start);
 
